homeassistant/components/sonos/media_player.py

```diff
@@ -326,53 +324,49 @@ def coordinator(self):
         """Return coordinator of this player."""
         return self._coordinator
 
-    def seen(self):
+    async def async_seen(self):
```

Comment:
Curious why making this `async` as it's not using any real async work and is called from both a sync and async context (which can just add the job).

Reply:
I realized that there are possible races, for example here:
```
        if self._seen_timer:
            self._seen_timer()
```

Reply:
Sorry not following, how does `async` solve that?  Sounds like you might need a lock?

Reply:
Being `async` is an implicit lock since running coroutines are not preempted.

Reply:
Right, but control is yielded back to the loop with the await on line 334, so there's no guarantee another instance of this couroutine won't get kicked off before this one is completed.  If it didn't have any `await` at all, then it would run without interruption.  If the code above LN334 is the sensitive part (and not below), then there's no issue or need to use a lock.

Reply:
Ah, I see what you mean now. I was thinking about "seen" racing with "unseen" which could then clear `self._seen_timer` between the above two statements.

You are technically correct that "seen" can still race with itself if it manages to take more than a full minute to complete so the next discovery event happens. I have reordered the sections to make that even less fatal.

Reply:
Be careful to mix sync/async. The reason is simple. If you wait to an executor and the executor is blocking to an event they are async - you will become a deadlock with more as 10boxes. That is only as remember

Reply:
@pvizeli I am definitely not liking this mix but it seems needed when `threading.Lock` is discouraged in executor threads.

I would probably be rewriting the third party library for async if not for the fact that Sonos has announced (but not shipped) a new LAN API. Oh, and the UDP discovery would not work with Windows.